File: InventarioApp/DataStore.swift
import SwiftUI

class DataStore: ObservableObject {
    // MARK: - Published State
    @Published var catalog: [CatalogProduct] = []
    @Published var tables: [String] = []
    @Published var activeTable: String = ""
    @Published var currentFecha: String = ""
    @Published var entries: [JornadaEntry] = []
    @Published var historial: [Jornada] = []
    @Published var isEditingHistorical = false
    @Published var showBackupReminder = false

    private var sessions: [String: SavedSession] = [:]
    private var saveCount: Int = 0

    // UserDefaults keys
    private let kCatalog   = "inv2_catalog"
    private let kTables    = "inv2_tables"
    private let kActive    = "inv2_active"
    private let kHist      = "inv2_hist"
    private let kSessions  = "inv2_sessions"
    private let kSaveCount = "inv2_save_count"

    var catalogIsEmpty: Bool { catalog.isEmpty }

    var totalImporte: Double {
        entries.reduce(0) { $0 + $1.importe }
    }

    var detalleImporte: [(String, Double)] {
        entries.compactMap { e in
            e.importe > 0 ? (e.nombre, e.importe) : nil
        }
    }

    var discrepancias: [(String, Double)] {
        entries.compactMap { e in
            if let d = e.discrepancia { return (e.nombre, d) }
            return nil
        }
    }

    var filteredHistorial: [Jornada] {
        historial
    }

    // MARK: - Init
    init() {
        loadAll()
    }

    // MARK: - Load All
    func loadAll() {
        catalog = load(kCatalog) ?? []
        tables = load(kTables) ?? []
        if tables.isEmpty { tables = ["Tabla 1"] }

        activeTable = UserDefaults.standard.string(forKey: kActive) ?? tables.first ?? "Tabla 1"
        if !tables.contains(activeTable) { activeTable = tables.first ?? "Tabla 1" }

        historial = load(kHist) ?? []
        sessions = load(kSessions) ?? [:]
        saveCount = UserDefaults.standard.integer(forKey: kSaveCount)

        loadCurrentSession()
    }

    // MARK: - Session Management
    func loadCurrentSession() {
        if let session = sessions[activeTable] {
            currentFecha = session.fecha
            entries = session.filas
        } else {
            currentFecha = Self.todayString()
            entries = catalog.map { p in
                JornadaEntry(nombre: p.nombre, precio: String(p.precio))
            }
        }
    }

    func saveCurrentSession() {
        sessions[activeTable] = SavedSession(fecha: currentFecha, filas: entries)
        save(sessions, kSessions)
    }

    // MARK: - Catalog
    func saveCatalog() {
        save(catalog, kCatalog)
    }

    // MARK: - Table Management
    func switchTable(to name: String) {
        saveCurrentSession()
        activeTable = name
        UserDefaults.standard.set(name, forKey: kActive)
        isEditingHistorical = false
        loadCurrentSession()
    }

    func addTable(name: String) -> Bool {
        let n = name.trimmingCharacters(in: .whitespaces)
        guard !n.isEmpty, !tables.contains(n) else { return false }
        saveCurrentSession()
        tables.append(n)
        activeTable = n
        saveTables()
        isEditingHistorical = false
        loadCurrentSession()
        return true
    }

    func renameTable(newName: String) -> Bool {
        let n = newName.trimmingCharacters(in: .whitespaces)
        guard !n.isEmpty, n != activeTable, !tables.contains(n) else { return false }

        if let session = sessions[activeTable] {
            sessions[n] = session
            sessions.removeValue(forKey: activeTable)
        }

        historial = historial.map { j in
            if j.tabla == activeTable {
                return Jornada(id: j.id, fecha: j.fecha, tabla: n,
                               totalImporte: j.totalImporte, filas: j.filas)
            }
            return j
        }
        saveHistorial()

        if let idx = tables.firstIndex(of: activeTable) {
            tables[idx] = n
        }
        activeTable = n
        saveTables()
        save(sessions, kSessions)
        return true
    }

    func deleteTable() -> Bool {
        guard tables.count > 1 else { return false }
        sessions.removeValue(forKey: activeTable)
        tables.removeAll { $0 == activeTable }
        activeTable = tables.first ?? "Tabla 1"
        saveTables()
        save(sessions, kSessions)
        isEditingHistorical = false
        loadCurrentSession()
        return true
    }

    private func saveTables() {
        save(tables, kTables)
        UserDefaults.standard.set(activeTable, forKey: kActive)
    }

    // MARK: - Entry Management
    func addEntry(_ entry: JornadaEntry? = nil) {
        entries.append(entry ?? JornadaEntry())
        saveCurrentSession()
    }

    func removeEntry(id: UUID) {
        entries.removeAll { $0.id == id }
        saveCurrentSession()
    }

    func moveEntry(from source: IndexSet, to destination: Int) {
        entries.move(fromOffsets: source, toOffset: destination)
        saveCurrentSession()
    }

    // MARK: - Guardar Día
    func guardarDia() -> String? {
        guard !currentFecha.isEmpty else {
            return "Indica la fecha de la jornada."
        }

        let total = ((totalImporte * 100).rounded() / 100)
        let jornada = Jornada(fecha: currentFecha, tabla: activeTable,
                              totalImporte: total, filas: entries)

        historial.removeAll { $0.fecha == currentFecha && $0.tabla == activeTable }
        historial.insert(jornada, at: 0)
        historial.sort { ($0.fecha, $0.tabla) > ($1.fecha, $1.tabla) }

        saveHistorial()
        isEditingHistorical = false

        saveCount += 1
        UserDefaults.standard.set(saveCount, forKey: kSaveCount)
        if saveCount % 7 == 0 {
            showBackupReminder = true
        }

        return nil
    }

    func saveHistorial() {
        save(historial, kHist)
    }

    // MARK: - History Actions
    func editJornada(_ jornada: Jornada) {
        saveCurrentSession()

        if jornada.tabla != activeTable {
            if !tables.contains(jornada.tabla) {
                tables.append(jornada.tabla)
                saveTables()
            }
            activeTable = jornada.tabla
            UserDefaults.standard.set(activeTable, forKey: kActive)
        }

        currentFecha = jornada.fecha
        entries = jornada.filas
        isEditingHistorical = true
        saveCurrentSession()
    }

    func deleteJornada(_ jornada: Jornada) {
        historial.removeAll { $0.id == jornada.id }
        saveHistorial()
    }

    func cancelarEdicion() {
        isEditingHistorical = false
    }

    // MARK: - Limpiar
    func limpiar() {
        currentFecha = Self.todayString()
        entries = catalog.map { p in
            JornadaEntry(nombre: p.nombre, precio: String(p.precio))
        }
        sessions.removeValue(forKey: activeTable)
        save(sessions, kSessions)
        isEditingHistorical = false
    }

    // MARK: - CSV Generation
    func generateCSV() -> String {
        var csv = "Tabla,Fecha,Producto,Inicial,Venta,Precio,Importe,Final\n"
        for e in entries {
            let imp = String(format: "%.2f", e.importe)
            csv += "\"\(activeTable)\",\(currentFecha),\"\(e.nombre)\",\(e.inicial),\(e.venta),\(e.precio),\(imp),\(e.finalVal)\n"
        }
        let total = String(format: "%.2f", totalImporte)
        csv += "\n,,,,,Total,\(total),\n"
        return csv
    }

    // MARK: - Helpers
    static func todayString() -> String {
        let df = DateFormatter()
        df.dateFormat = "yyyy-MM-dd"
        return df.string(from: Date())
    }

    static func dateFromString(_ s: String) -> Date {
        let df = DateFormatter()
        df.dateFormat = "yyyy-MM-dd"
        return df.date(from: s) ?? Date()
    }

    static func stringFromDate(_ d: Date) -> String {
        let df = DateFormatter()
        df.dateFormat = "yyyy-MM-dd"
        return df.string(from: d)
    }

    private func load<T: Decodable>(_ key: String) -> T? {
        guard let data = UserDefaults.standard.data(forKey: key) else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }

    private func save<T: Encodable>(_ value: T, _ key: String) {
        if let data = try? JSONEncoder().encode(value) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }
}

File: android/app/src/main/java/com/inventario/app/data/DataStore.kt
package com.inventario.app.data

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.inventario.app.model.*
import java.text.SimpleDateFormat
import java.util.*

class InventarioDataStore(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("inventario", Context.MODE_PRIVATE)
    private val gson = Gson()

    // Keys
    private val kCatalog = "inv2_catalog"
    private val kTables = "inv2_tables"
    private val kActive = "inv2_active"
    private val kHist = "inv2_hist"
    private val kSessions = "inv2_sessions"
    private val kSaveCount = "inv2_save_count"

    // State
    var catalog = mutableStateListOf<CatalogProduct>()
    var tables = mutableStateListOf<String>()
    var activeTable by mutableStateOf("")
    var currentFecha by mutableStateOf(todayString())
    var entries = mutableStateListOf<JornadaEntry>()
    var historial = mutableStateListOf<Jornada>()
    var isEditingHistorical by mutableStateOf(false)
    var showBackupReminder by mutableStateOf(false)

    private var sessions = mutableMapOf<String, SavedSession>()
    private var saveCount = 0

    val catalogIsEmpty: Boolean get() = catalog.isEmpty()

    val totalImporte: Double get() = entries.sumOf { it.importe }

    val detalleImporte: List<Pair<String, Double>>
        get() = entries.filter { it.importe > 0 }.map { it.nombre to it.importe }

    val discrepancias: List<Pair<String, Double>>
        get() = entries.mapNotNull { e -> e.discrepancia?.let { e.nombre to it } }

    init {
        loadAll()
    }

    fun loadAll() {
        catalog.clear()
        catalog.addAll(loadList(kCatalog) ?: emptyList())

        tables.clear()
        val loadedTables: List<String> = loadList(kTables) ?: listOf("Tabla 1")
        tables.addAll(if (loadedTables.isEmpty()) listOf("Tabla 1") else loadedTables)

        activeTable = prefs.getString(kActive, tables.firstOrNull() ?: "Tabla 1") ?: "Tabla 1"
        if (activeTable !in tables) activeTable = tables.firstOrNull() ?: "Tabla 1"

        historial.clear()
        historial.addAll(loadList(kHist) ?: emptyList())

        sessions = loadMap(kSessions) ?: mutableMapOf()
        saveCount = prefs.getInt(kSaveCount, 0)

        loadCurrentSession()
    }

    fun loadCurrentSession() {
        val session = sessions[activeTable]
        entries.clear()
        if (session != null) {
            currentFecha = session.fecha
            entries.addAll(session.filas)
        } else {
            currentFecha = todayString()
            entries.addAll(catalog.map { JornadaEntry(nombre = it.nombre, precio = it.precio.toString()) })
        }
    }

    fun saveCurrentSession() {
        sessions[activeTable] = SavedSession(fecha = currentFecha, filas = entries.toList())
        saveMap(sessions, kSessions)
    }

    fun saveCatalog() {
        saveList(catalog.toList(), kCatalog)
    }

    // Table management
    fun switchTable(name: String) {
        saveCurrentSession()
        activeTable = name
        prefs.edit().putString(kActive, name).apply()
        isEditingHistorical = false
        loadCurrentSession()
    }

    fun addTable(name: String): Boolean {
        val n = name.trim()
        if (n.isEmpty() || n in tables) return false
        saveCurrentSession()
        tables.add(n)
        activeTable = n
        saveTables()
        isEditingHistorical = false
        loadCurrentSession()
        return true
    }

    fun renameTable(newName: String): Boolean {
        val n = newName.trim()
        if (n.isEmpty() || n == activeTable || n in tables) return false

        sessions[n] = sessions.remove(activeTable) ?: SavedSession(todayString(), emptyList())

        val updated = historial.map { j ->
            if (j.tabla == activeTable) j.copy(tabla = n) else j
        }
        historial.clear()
        historial.addAll(updated)
        saveHistorial()

        val idx = tables.indexOf(activeTable)
        if (idx >= 0) tables[idx] = n
        activeTable = n
        saveTables()
        saveMap(sessions, kSessions)
        return true
    }

    fun deleteTable(): Boolean {
        if (tables.size <= 1) return false
        sessions.remove(activeTable)
        tables.remove(activeTable)
        activeTable = tables.firstOrNull() ?: "Tabla 1"
        saveTables()
        saveMap(sessions, kSessions)
        isEditingHistorical = false
        loadCurrentSession()
        return true
    }

    private fun saveTables() {
        saveList(tables.toList(), kTables)
        prefs.edit().putString(kActive, activeTable).apply()
    }

    // Entry management
    fun addEntry(entry: JornadaEntry = JornadaEntry()) {
        entries.add(entry)
        saveCurrentSession()
    }

    fun removeEntry(id: String) {
        entries.removeAll { it.id == id }
        saveCurrentSession()
    }

    fun updateEntry(index: Int, entry: JornadaEntry) {
        if (index in entries.indices) {
            entries[index] = entry
            saveCurrentSession()
        }
    }

    // Guardar día
    fun guardarDia(): String? {
        if (currentFecha.isEmpty()) return "Indica la fecha de la jornada."

        val total = Math.round(totalImporte * 100.0) / 100.0
        val jornada = Jornada(
            fecha = currentFecha,
            tabla = activeTable,
            totalImporte = total,
            filas = entries.toList()
        )

        historial.removeAll { it.fecha == currentFecha && it.tabla == activeTable }
        historial.add(0, jornada)
        historial.sortWith(compareByDescending<Jornada> { it.fecha }.thenBy { it.tabla })
        saveHistorial()
        isEditingHistorical = false

        saveCount++
        prefs.edit().putInt(kSaveCount, saveCount).apply()
        if (saveCount % 7 == 0) showBackupReminder = true

        return null
    }

    fun saveHistorial() {
        saveList(historial.toList(), kHist)
    }

    // History actions
    fun editJornada(jornada: Jornada) {
        saveCurrentSession()
        if (jornada.tabla != activeTable) {
            if (jornada.tabla !in tables) {
                tables.add(jornada.tabla)
                saveTables()
            }
            activeTable = jornada.tabla
            prefs.edit().putString(kActive, activeTable).apply()
        }
        currentFecha = jornada.fecha
        entries.clear()
        entries.addAll(jornada.filas)
        isEditingHistorical = true
        saveCurrentSession()
    }

    fun deleteJornada(jornada: Jornada) {
        historial.removeAll { it.id == jornada.id }
        saveHistorial()
    }

    fun cancelarEdicion() {
        isEditingHistorical = false
    }

    fun limpiar() {
        currentFecha = todayString()
        entries.clear()
        entries.addAll(catalog.map { JornadaEntry(nombre = it.nombre, precio = it.precio.toString()) })
        sessions.remove(activeTable)
        saveMap(sessions, kSessions)
        isEditingHistorical = false
    }

    // CSV
    fun generateCSV(): String {
        val sb = StringBuilder("Tabla,Fecha,Producto,Inicial,Venta,Precio,Importe,Final\n")
        for (e in entries) {
            val imp = String.format("%.2f", e.importe)
            sb.append("\"$activeTable\",$currentFecha,\"${e.nombre}\",${e.inicial},${e.venta},${e.precio},$imp,${e.finalVal}\n")
        }
        sb.append("\n,,,,,Total,${String.format("%.2f", totalImporte)},\n")
        return sb.toString()
    }

    // Helpers
    companion object {
        fun todayString(): String {
            val sdf = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
            return sdf.format(Date())
        }
    }

    private inline fun <reified T> loadList(key: String): List<T>? {
        val json = prefs.getString(key, null) ?: return null
        return try {
            gson.fromJson(json, TypeToken.getParameterized(List::class.java, T::class.java).type)
        } catch (e: Exception) { null }
    }

    private fun <T> saveList(list: List<T>, key: String) {
        prefs.edit().putString(key, gson.toJson(list)).apply()
    }

    private fun loadMap(key: String): MutableMap<String, SavedSession>? {
        val json = prefs.getString(key, null) ?: return null
        return try {
            gson.fromJson(json, object : TypeToken<MutableMap<String, SavedSession>>() {}.type)
        } catch (e: Exception) { null }
    }

    private fun saveMap(map: Map<String, SavedSession>, key: String) {
        prefs.edit().putString(key, gson.toJson(map)).apply()
    }
}
